vllm/attention/backends/dual_chunk_flash_attn.py

```diff
@@ -0,0 +1,877 @@
+                "Native Dynamic Chunk Attention does not support window_size")
+
+        cu_seqlens_q_cpu = cu_seqlens_q.cpu().tolist()
+        cu_seqlens_k_cpu = cu_seqlens_k.cpu().tolist()
+
+        all_outputs = []
+        for i in range(0, len(cu_seqlens_q_cpu) - 1):
+            qs = cu_seqlens_q_cpu[i]
+            qe = cu_seqlens_q_cpu[i:i + 2][-1]
+            ks = cu_seqlens_k_cpu[i]
+            ke = cu_seqlens_k_cpu[i:i + 2][-1]
+
+            current_q = q[qs:qe]
+            current_q_succ = q_succ[qs:qe]
+            current_q_inter = q_inter[qs:qe]
+            if block_table is None:
+                current_k = k[ks:ke]
+                current_v = v[ks:ke]
+                current_block_table = None
+                current_prefill_original_seq_lens_tensor = (
+                    prefill_original_seq_lens_tensor[i:i + 1])
+            else:
+                current_block_table = block_table[i:i + 1]
+                current_prefill_original_seq_lens_tensor = (
+                    prefill_original_seq_lens_tensor[i:i + 1])
+                current_k = k
+                current_v = v
+
+            if current_q.shape[0] == 0:
+                continue
+            if current_k.shape[0] == 0:
+                all_outputs.append(
+                    torch.zeros(
+                        (current_q.shape[0], current_q.shape[1], v.shape[2]),
+                        device=q.device,
+                        dtype=q.dtype,
+                    ))
+                continue
+
+            current_output = self._bruteforce_dynamic_chunk_flash_attn_func(
+                current_q,
+                current_q_succ,
+                current_q_inter,
+                current_k,
+                current_v,
+                current_block_table,
+                softmax_scale,
+                chunk_size,
+                local_size,
+                original_max_position_embeddings,
+                current_prefill_original_seq_lens_tensor,
+                ke - ks,
+            )
+            all_outputs.append(current_output)
+
+        return torch.cat(all_outputs, dim=0)
+
+    def _bruteforce_dynamic_chunk_pageattention_forward_decode(
+        self,
+        query: torch.Tensor,
+        query_succ: torch.Tensor,
+        query_inter: torch.Tensor,
+        key_cache: torch.Tensor,
+        value_cache: torch.Tensor,
+        block_table: torch.Tensor,
+        cache_seqlens: torch.Tensor,
+        softmax_scale: float,
+        causal: bool,
+        alibi_slopes: Optional[torch.Tensor],
+        chunk_size: int,
+        local_size: int,
+        original_max_position_embeddings: int,
+    ):
+        assert causal
+        batch_size = block_table.shape[0]
+        block_size = value_cache.shape[1]
+        chunk_len = chunk_size - local_size
+        if chunk_len % block_size != 0:
+            raise ValueError("chunk_len must be divisible by block_size.")
+        chunk_num_curr = (cache_seqlens - 1) // chunk_len
```

Comment:
This line makes the actual length of `intra-chunk` smaller than `chunk_len` in the variable length situation. It differs from the algorithm in the prefill stage with a fixed length of `chunk_len`. Could you explain the reason for this difference?

Reply:
@rejoicesyc The implementation is consistent with the original code (https://github.com/HKUNLP/ChunkLlama/blob/8a28f1464b2a5def03eb07e8d91f5bf4d00f667d/chunkllama_attn_replace.py#L167). See L167, L171, L173, L197.

Since the sequence length may not be divisible by ``chunk_len``, there has to be a chunk that has a smaller length. In DCA algorithm, the last chunk (i.e., the intra chunk) may have ``length <= chunk_len`` while the previous chunk has a fixed length of ``chunk_len``.